### PR TITLE
SPLICE-1603 support sample stats:

### PR DESCRIPTION
**Pull request by yxia92**
1. add system procedure collect_table_sample_statistics;
2. collect stats based on the sample percentage specified;
3. introduce two new columns: statsType, sampleFraction to the system table sys.systablestats, and to the system view sys.systablestatistics
4. adjust estimation logic to exptrapolate/scale based on the sample fraction
5. add upgrade logic for the addition of two new columns to the dictionary tables,
   and to update the view defintiion for systablestatistics
6. add syntax support of sample stats for analyze command

### CODE REVIEW

**Comment by yxia92** [comment]
While preparing the fix branch for 2.5, I noticed an issue with stats collection for external table, so updated the pull request with a fix. 

Here is the details:

The following code change in StatisticsOperation.getDataSet() to pass "operationContext" instead of "null" could have a side effect for stats collection on external tables:
-                    statsDataSet = dsp.readTextFile(null, builder.getLocation(), builder.getDelimited(), null, builder.getColumnPositionMap(), null, null, null, builder.getTemplate());
+                    statsDataSet = dsp.readTextFile(null, builder.getLocation(), builder.getDelimited(), null, builder.getColumnPositionMap(), operationContext, null, null, builder.getTemplate());

Here is an example:
CREATE EXTERNAL TABLE et1(
   a1 INT, b1 int, c1 int)
   PARTITIONED BY (a1)
   ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' ESCAPED BY '\\'
   LINES TERMINATED BY '\\n'
   STORED AS TEXTFILE
   LOCATION '/Users/yxia/Documents/work/SPLICE-1603/et1';

insert into et1 values (1,1,1);
insert into et1 values (2,2,2);
insert into et1 values (3,3,3);
insert into et1 values (4,4,4);
insert into et1 values (5,5,5);
insert into et1 values (6,6,6);
insert into et1 values (7,7,7);
insert into et1 values (8,8,8);
insert into et1 values (9,9,9);
insert into et1 values (10,10,10);

We get the following output for analyze. Note, the value for  "partition" is "--".
splice> analyze table et1;
schemaName                                                                                                                                                                                                                                                      |tableName                                                                                                                                                                                                                                                       |partition                                                                                                                                                                                                                                                       |rowsCollec&|partitionSize       |partitionCount      |statsType  |sampleFraction
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
TEST                                                                                                                                                                                                                                                            |ET1                                                                                                                                                                                                                                                             |--                                                                                                                                                                                                                                                              |4          |48                  |1                   |0          |0.0
TEST                                                                                                                                                                                                                                                            |ET1                                                                                                                                                                                                                                                             |--                                                                                                                                                                                                                                                              |4          |48                  |1                   |0          |0.0
TEST                                                                                                                                                                                                                                                            |ET1                                                                                                                                                                                                                                                             |--                                                                                                                                                                                                                                                              |2          |24                  |1                   |0          |0.0

3 rows selected 

The expected value for "partition" is 0, 1, 2 for the three rows respectively.

To avoid this side effect, make the change to keep the original null parameter, and pass useSample and sampleFraction as two new parameters to readTextFile().

Similar change is made to readParquetFile() and readORCFile().

Here is the list of files touched, they are mainly interface change to add the two new parameters:

	modified:   hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSetProcessor.java
	modified:   hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java
	modified:   splice_machine/src/main/java/com/splicemachine/derby/stream/control/ControlDataSetProcessor.java
	modified:   splice_machine/src/main/java/com/splicemachine/derby/stream/iapi/DataSetProcessor.java
	modified:   splice_machine/src/main/java/com/splicemachine/derby/stream/utils/ForwardingDataSetProcessor.java
	modified:   splice_machine/src/main/java/com/splicemachine/derby/utils/StatisticsOperation.java

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSetProcessor.java [hunk 1]
```diff
@@ -327,9 +307,16 @@ public Partitioner getPartitioner(DataSet<LocatedRow> dataSet, ExecRow template,
...
             }
             table = processExternalDataset(table,baseColumnMap,qualifiers,probeValue);
-            return new SparkDataSet(table
-                    .rdd().toJavaRDD()
-                    .map(new RowToLocatedRowFunction(context,execRow)));
+
+            if (useSample) {
+                return new SparkDataSet(table
+                        .rdd().toJavaRDD()
+                        .map(new RowToLocatedRowFunction(context, execRow)).sample(false, sampleFraction));
```
Comment (wtang16) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSetProcessor.java]: Something Yi needs to change for this PR or a follow-up PR?

Comment (yxia92) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSetProcessor.java]: Thanks @dgomezferro . I can update the code in this PR. I'll update the code after collecting all the comments.

**Comment by yxia92** [comment]
The sampling may add a little overhead, but the reduction in the number of rows to feed to stats collection logic should compensate for it. In SPLICE-1603, I have the elapse time comparison on collecting sample stats on the Orders table in TPCH100G at different sample fraction:
collect_stats.pdf <https://splice.atlassian.net/secure/attachment/17601/17601_collect_stats.pdf>
The data point at sample fraction 1 is the regular stats.
 <https://splice.atlassian.net/secure/attachment/17601/collect_stats.pdf> <https://splice.atlassian.net/secure/attachment/17601/collect_stats.pdf>
 <https://splice.atlassian.net/secure/attachment/17601/collect_stats.pdf>

**Comment by snowspeeder** [comment]
+1

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSetProcessor.java (hunk 1 above)
Comment (yxia92) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkDataSetProcessor.java]: @dgomezferro I've updated the PR to move sample() before map(). Same change is made to the PR for 2.5 too.

**Comment by wtang16** [comment]
+1

**Pull request merged**